File: code-orig/cpp/src/particle.cpp
#include "../include/particle.h"

Particle::Particle(Vec2d pos, Vec2d vel, double radius, int idx) : pos(pos), vel(vel), radius(radius), idx(idx) {}

void Particle::setvel(Vec2d vel) {
    this->vel = vel;
}
void Particle::setpos(Vec2d pos) {
    this->pos = pos;
}

bool Particle::operator==(const Particle& other) const {
    return idx == other.idx;
}

void Particle::update() {
    pos = pos + vel * dt;
    
    if(pos.x < 0) {
        pos.x = 0;
        vel.x *= -1;
    }
    if(pos.x > X_SIZE) {
        pos.x = X_SIZE;
        vel.x *= -1;
    }
    if(pos.y < 0) {
        pos.y = 0;
        vel.y *= -1;
    }
    if(pos.y > Y_SIZE) {
            pos.y = Y_SIZE;
            vel.y *= -1;
    }
    }

double distance(const Particle& lhs, const Particle& rhs) {
    return (lhs.pos - rhs.pos).magnitude();
}

File: code-orig/rust/src/particle.rs
use crate::vec2d::Vec2d; // Assuming vec2d module is in the same crate
use crate::config::DT;

#[derive(Debug, PartialEq, Clone)]
pub struct Particle {
    pub pos: Vec2d,
    pub vel: Vec2d,
    pub radius: f64,
    pub idx: i32,
}

impl Particle {
    pub fn new(pos: Vec2d, vel: Vec2d, radius: f64, idx: i32) -> Self {
        Particle { pos, vel, radius, idx }
    }

    pub fn custom_clone(&self) -> Self {
        Particle {
            pos: self.pos,
            vel: self.vel,
            radius: self.radius,
            idx: self.idx,
        }
    }

    pub fn set_vel(&mut self, vel: Vec2d) {
        self.vel = vel;
    }

    pub fn set_pos(&mut self, pos: Vec2d) {
        self.pos = pos;
    }

    pub fn update(&mut self) {
        self.pos = self.pos + self.vel * DT;
    }
}